### ISSUE TITLE
Support more of abstract array interface

### ISSUE THREAD

**Issue by MilesCranmer** [JuliaPy/PythonCall.jl#462]
Currently `similar` will return an array of a different type:

```python
from juliacall import Main as jl
import numpy as np

x = np.random.randn(5)
y = jl.similar(x)

print(jl.typeof(x))
# PyArray{Float64, 1, true, true, Float64}
print(jl.typeof(y))
# Vector{Float64}
```

This can introduce some type instabilities in libraries due to the assumption that container type is preserved by `similar`.

I guess we just need the "optional methods" from here: https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array

**Comment by marcosdiezgarcia** [comment]
```julia
julia> using PythonCall

julia> Py(rand(10)).to_numpy().dtype
Python: dtype('float64')

julia> Py([[1,2,3], [4,5,6]]).to_numpy().dtype
Python: dtype('O')
```

I would expect the latter to be `dtype('int64')` to match Python:

```python
>>> import numpy
>>> a = numpy.array([[1,2,3], [4,5,6]])
>>> a.dtype
dtype('int64')
```

**Comment by marcosdiezgarcia** [comment]
Yes, of course, the `dtype` itself is an object. Specifically, it is an `Int64DType` as shown below. But I was referring to the types of the elements contained in the numpy array when I asked
> Is there any possible way to make the dtype be dtype('int64') instead of dtype('O')?

Regardless of how many levels of nesting, `dtype('int64')` is the value of the property `dtype` as shown below:

<img width="735" height="209" alt="Image" src="https://github.com/user-attachments/assets/ec3d59b0-4534-448c-a529-292070e5c322" />

Whereas in Julia, using PythonCall `to_numpy()`, it is clear that `dtype('int64')` is only output for `Vector{Int64}` not `Vector{Vector{Int64}}` nor `Vector{Vector{Vector{Int64}}}`:

<img width="801" height="216" alt="Image" src="https://github.com/user-attachments/assets/f2d26a1b-d0c7-41cb-a182-d22fb1a35dd6" />

To me this shows that the value of the property `.dtype` in numpy is defined based on the innermost elements in the array, whereas in Julia the value of `.dtype` upon using `to_numpy()` is not based on the innermost elements (i.e. 1, 2, 3, etc) but on the whole structure containing them (i.e `Vector{Int64}` for the 2nd array, and `Vector{Vector{Int64}}` for the 3rd array). (Perhaps this is what you were trying to explain above.) So it is clear that the behaviour of accessing the property `.dtype` is different in numpy compared to what julicall outputs from `to_numpy()`. I was expecting same behaviour.

**Comment by MilesCranmer** [comment]
Again, this is because numpy automatically converts to a single 2D array, whereas Julia leaves it as a vector of vectors. There is no bug here.

**Comment by marcosdiezgarcia** [comment]
I found a simple fix to `JlWrap/array.jl` ([line 370](https://github.com/JuliaPy/PythonCall.jl/blob/71666ca0a30eafee2456a191fa2490cdd0b0ed22/src/JlWrap/array.jl#L370)) in PythonCall 0.9.28:

```diff
try:
    import numpy
-   arr = numpy.array(arr, dtype=dtype)
+   arr = numpy.array(list(arr), dtype=dtype)
except ImportError:
    pass
```

Now I get what I was looking for:

```julia
julia> Py([1, 2, 3]).to_numpy().dtype
Python: dtype('int64')

julia> Py([[1, 2, 3], [4, 5, 6]]).to_numpy().dtype
Python: dtype('int64')

julia> Py([[[1, 2, 3], [4, 5, 6]]]).to_numpy().dtype
Python: dtype('int64')
```

This also works if the underlying elements are of other type like floats, e.g.:

```julia
julia> Py([[1.0,2.0,3.0], [4.1,5.1,6.1]]).to_numpy()
Python:
array([[1. , 2. , 3. ],
       [4.1, 5.1, 6.1]])

julia> Py([[1.0,2.0,3.0], [4.1,5.1,6.1]]).to_numpy().dtype
Python: dtype('float64')
```

**Comment by marcosdiezgarcia** [comment]
No worries. I'll raise a separate issue.
